### PR TITLE
C++: Include "phi reads" in `DataFlow::Node`

### PR DESCRIPTION
**Pull request by MathiasVP**
In https://github.com/github/codeql/pull/11198 Tom added phi nodes for uses in the shared SSA library, but we haven't yet reaped the benefit of those in C++ since we've been using the backwards compatibility wrappers he made. This PR makes use of those new phi nodes for uses. It should give us a small performance boost 🤞.

[The first commit](d845cd7233117c805c9ac105b246eab8ccc04aff) does the real work, and the [second commit](34f8d4d3ebe549cc77d5b1a369c2583c4dc87a36) fixes a bug exposed by the tests.

### CODE REVIEW

**Comment by MathiasVP** [comment]
Hmmm... I don't understand why we gain so many new results from this, but they look like TPs to me 😮. The two results on `openjdk-jdk` doesn't even involve global variables, so https://github.com/github/codeql/pull/12356/commits/d845cd7233117c805c9ac105b246eab8ccc04aff#diff-e96bb3af46f5cc3580be03e17e599c937b020235aa92af08b77b1b2697d37d3cR560-R564 can't even be attributed to that 🤔.

**Comment by jketema** [comment]
> Hmmm... I don't understand why we gain so many new results from this, but they look like TPs to me 😮. The two results on `openjdk-jdk` doesn't even involve global variables, so [d845cd7#diff-e96bb3af46f5cc3580be03e17e599c937b020235aa92af08b77b1b2697d37d3cR560-R564](https://github.com/github/codeql/commit/d845cd7233117c805c9ac105b246eab8ccc04aff#diff-e96bb3af46f5cc3580be03e17e599c937b020235aa92af08b77b1b2697d37d3cR560-R564) can't even be attributed to that 🤔.

You also fixed a bug. Does that have something to do with it?

**Comment by MathiasVP** [comment]
> You also fixed a bug. Does that have something to do with it?

Good point. I had the exact same thought, so I actually made a PR with just that fix [here](https://github.com/github/codeql/pull/12353) and ran a DCA on that, but as you can see from the DCA run, that did _not_ actually produce any new results.

**Comment by MathiasVP** [comment]
I _may_ understand part of the issue now. I've found the step that's responsible for the new results on `cpp/path-injection` [here](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/hotspot/share/utilities/vmError.cpp#L1328): We now step from the output argument `buffer` to a new [phi-read node](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/hotspot/share/utilities/vmError.cpp#L1413) and further to the sink [here](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/hotspot/share/utilities/vmError.cpp#L1427).

Notice that `buffer` is a static variable. This means we don't actually have proper dataflow support for it right now (static locals are missing an initial definition). Before this PR, there would be no step from the [argument node](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/hotspot/share/utilities/vmError.cpp#L1328) to the [sink](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/hotspot/share/utilities/vmError.cpp#L1427), but on this PR there's a new SSA phi-read node for the read of [the argument]((https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/hotspot/share/utilities/vmError.cpp#L1413)). This new SSA phi-read node _is_ a definition (because phi nodes are definitions), so the `buffer` variable now has a definition, and thus we get SSA flow that reach the sink 🎉.

Now, am I happy with the fact that this is how we're getting the flow? No, not really 😂. But we would have gotten this result anyway once we implement proper support for static locals, so I don't really mind the fact that we're getting this result now. This explains the two new results on `cpp/path-injection`. I'll continue investigating the other results as well and see if we get those for similar reasons.

**Comment by MathiasVP** [comment]
I'm seeing the exact same situation on the new `cpp/uncontrolled-process-operation` result where [this](https://github.com/kamailio/kamailio/blob/b02c247023ea3ea0ef9753efdb04aff7d5d3bbb4/src/modules/exec/exec.c#L153) static local variable receives flow from an out argument.

**Comment by MathiasVP** [comment]
I manage to reproduce the above situation in a QL test 🎉. I've rebased the PR to make it follow the usual "add test", "write QL", "accept tests" workflow. No actual QL code has changed in the force-push!

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll [hunk 1]
```diff
@@ -550,11 +550,18 @@ predicate adjacentDefRead(DefOrUse defOrUse1, UseOrPhi use) {
...
+    exists(IRBlock bb2, int i2 |
+      adjacentDefReadExt(_, pragma[only_bind_into](bb1), pragma[only_bind_into](i1),
+        pragma[only_bind_into](bb2), pragma[only_bind_into](i2)) and
+      useOrPhi.asDefOrUse().hasIndexInBlock(bb2, i2, v)
+    )
+    or
+    exists(PhiNode phi |
+      lastRefRedefExt(_, bb1, i1, phi) and
+      useOrPhi.asPhi() = phi and
+      phi.getSourceVariable() = pragma[only_bind_into](v)
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:581]: PR mostly looks sensible. This is the only part I have trouble wrapping my head around. What does this do?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:581]: `lastRefRedefExt(_, bb1, i1, phi)` says that the last reference to the variable associated with `phi` was at index `i1` in basic block `bb1`. This means that `(bb1, i1)` is an input to `phi`. Does that help?

It's basically the same pattern as [here](https://github.com/github/codeql/blob/mathiasvp/replace-ast-with-ir-use-usedataflow/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L540). We should probably factor that out into a common predicate in another PR.

**Review by jketema — APPROVED** [review]

**Pull request merged**